[OpenMP][Flang] Emit default declare mappers implicitly for derived types

TIFitis opened the pull request:
This patch adds support to emit default declare mappers for implicit mapping of derived types when not supplied by user. This especially helps tackle mapping of allocatables of derived types.

TIFitis committed:
[OpenMP][Flang] Emit default declare mappers implicitly for derived types

This patch adds support to emit default declare mappers for implicit mapping of derived types when not supplied by user. This especially helps tackle mapping of allocatables of derived types.
This supports nested derived types as well.

skatrak reviewed (commented):
Thank you Akash for this work, and excuse the delay!

TIFitis commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2348,6 +2348,122 @@ genSingleOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
 }
 
+static mlir::FlatSymbolRefAttr
+genImplicitDefaultDeclareMapper(lower::AbstractConverter &converter,
```
It shouldn't be used with use_device_ptr and other such clauses as they don't support a declare mapper.

At the moment it's only used when the user hasn't made any effort to map a derived type. This can't be used in tangent with explicit map clause as map-type composition can cause divergence from the user's intention.

This is only mean't to help map derived types, especially, allocatables of derived types and nested derived types when no other mapping information is available.

TIFitis commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2348,6 +2348,122 @@ genSingleOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+    llvm::SmallVector<mlir::Value> bounds;
+    genBoundsOps(ref, bounds);
+    mlir::Value mapOp = createMapInfoOp(
+        firOpBuilder, loc, ref, /*varPtrPtr=*/mlir::Value{}, "", bounds,
```
The "Symbol" here is the `auto mapperArg = region.addArgument(firOpBuilder.getRefType(recordType), loc);` from above.

So there's no symname available for use here.

TIFitis commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2348,6 +2348,122 @@ genSingleOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
+  const auto getFieldRef =
+      [&](mlir::Value rec,
+          unsigned index) -> std::tuple<mlir::Value, mlir::Type> {
+    auto recType = mlir::dyn_cast<fir::RecordType>(
```
You're right we should be reusing `memberName` and `memberType` from the callsite, thanks for catching this.

skatrak reviewed (approved):
Thank you Akash for addressing my comments, and please excuse the delay taking a second look! LGTM.

agozillon reviewed (approved):
Hey Akash, this LGTM.

However, could you please add a test to the map-types-and-sizes.f90 test or create an equivalent for the mapper work! Would be good to keep track of what we expect the lowered LLVM-IR to be.

I also don't believe this PR handles more complicated mapping cases like below (please do correct me if I am wrong):

```
program main
    type :: bottom_layer
      integer(4), allocatable :: array_j3(:)
    end type bottom_layer

    type :: middle_layer
     type(bottom_layer), allocatable :: nest
     real(4), allocatable :: array_j2(:)
    end type middle_layer

    type :: top_layer
    integer, allocatable :: array_j(:)
    type(middle_layer), allocatable :: nested
    end type top_layer

    type(top_layer), allocatable :: top_dtype

    allocate(top_dtype)
    allocate(top_dtype%nested)
    allocate(top_dtype%nested%nest)
    allocate(top_dtype%array_j(10))
    allocate(top_dtype%nested%array_j2(10))
    allocate(top_dtype%nested%nest%array_j3(10))

!$omp target
  top_dtype%array_j(1) = 10
  top_dtype%nested%array_j2(1) = 20
  top_dtype%nested%nest%array_j3(1) = 30
!$omp end target

  print *, top_dtype%nested%nest%array_j3(1)
  print *, top_dtype%nested%array_j2(1)
  print *, top_dtype%array_j(1)

end program main

```

How difficult would it be to extend this work to handle the above cases? As I think that'd be the ideal next step for this work!

A really nice end goal for this work would be to handle slightly more complicated case than even the above with arbitrary nestings of array of allocatable derived types with allocatable types (arrays, scalars etc.)!

TIFitis committed:
Fix build error.

TIFitis committed:
Code cleanup.

TIFitis committed:
Add copilot suggestions.

agozillon commented on flang/lib/Lower/OpenMP/ClauseProcessor.cpp:
```diff
@@ -1290,13 +1336,28 @@ void ClauseProcessor::processMapObjects(
+            (objectTypeSpec &&
+             requiresImplicitDefaultDeclareMapper(*objectTypeSpec));
+        bool containsDelete = (mapTypeBits & mlir::omp::ClauseMapFlags::del) !=
+                              mlir::omp::ClauseMapFlags::none;
```
Perhaps a dumb question/thought, wouldn't we want to still emit in these cases? As we've implicitly allocated, I imagine implicit deallocation should perhaps still be done to prevent a user accidentally leaving bits and pieces on device, I could be wrong though!

TIFitis replied:
Sorry, I've fixed this.

agozillon commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2609,6 +2597,41 @@ genTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
               firOpBuilder, converter, defaultMaps, eleType, loc, sym);
 
+      mlir::FlatSymbolRefAttr mapperId;
+      if (defaultMaps.empty()) {
```
Might be good in the future to integrate this a little more elegantly with defaultmaps! As I imagine we don't want to prevent implicit mapping just because someone has specified a non-related implicit default map type (e.g. defaultmap(scalar: tofrom), and we might also want to take into account the user specified default mapping for either allocatable/pointer/aggregates, although not too sure where the specification stands on where each mapping would fall into or if they'd all fall under aggregate :-) 

But not the biggest deal at the moment for this PR, perhaps a good idea to add a TODO for it here though.

TIFitis replied:
Right now we bypass mapper generation whenever defaultMaps isn’t empty because we don’t know whether the user explicitly asked for defaultmap behaviour for aggregates/pointers.

I've added a TODO for this.

agozillon commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -537,6 +681,12 @@ void insertChildMapInfoIntoParent(
+      // Once explicit members are attached to a parent map, do not also invoke
+      // a declare mapper on it, otherwise the mapper would remap the same
+      // components leading to duplicate mappings at runtime.
+      if (!indices.second.memberMap.empty() && mapOp.getMapperIdAttr())
```
Perhaps a dumb question, would this prevent mapping of `scalar` in the below example (excuse the pseudo code!):

   type :: dtype2
    integer, allocatable :: scalar
   end type :: dtype2

    type :: dtype
    real(4) :: i 
    dtype2 :: dtype2_v
    end type dtype

    dtype :: dtype_v 

!$omp target map(dtype_v%i, dtype_2_v)
!$omp end target

As I think we'd still want to map it in these scenarios as far as I'm aware at least, Michael Klemm may know better.

agozillon replied:
Ah, makes sense, thank you for explaining!

agozillon reviewed (commented):
The PR LGTM so far, just a few questions and a NIT or two. And a request for one or two offload tests in check-offload if at all possible, just so we can see if there's any runtime regressions.

I was also wondering if this work would allow us to remove this segment of code: https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp#L948

I did test removing it and keeping just your PR in with the following example:

program main
implicit none
    integer:: i

    type :: bottom_layer
    real(4) :: i
    integer, allocatable :: scalar_i 
    integer(4) :: array_i(10)
    integer, allocatable :: array_k(:)
    integer(4) :: k
    end type bottom_layer
    
    type :: one_layer
    real(4) :: i
    integer, allocatable :: scalar 
    integer(4) :: array_i(10)
    real(4) :: j
    integer, allocatable :: array_j(:)
    integer(4) :: k
    type(bottom_layer) :: nest
    end type one_layer

    type(one_layer) :: one_l

    allocate(one_l%nest%array_k(10))
    allocate(one_l%array_j(10))
    allocate(one_l%nest%scalar_i) 
    allocate(one_l%scalar)

    do i = 1, 10
        one_l%nest%array_i(i) = i
    end do

    one_l%nest%scalar_i = 50
    one_l%scalar = 25

    !$omp target
    do i = 1, 10
        one_l%nest%array_k(i) = one_l%nest%array_i(i) + i + one_l%nest%scalar_i + one_l%scalar
        one_l%array_j(i) = one_l%nest%array_k(i)
    end do

    !$omp end target

print *, one_l%nest%array_k
print *, one_l%nest%array_i
print *, one_l%array_j
end program

Unfortunately, it seems that the MapInfoFinalization segment is still required for it, as there's unfortunately no declare mappers generated in this case (from what I could see when testing at least). Is this case something we could perhaps support with the declare mapper approach so we could unify the handling of implicit derived type member mapping? Not sure if it's a trivial extension or something that'd require a fair bit of legwork, if it's the latter we can likely leave it for now :-)

TIFitis commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -61,6 +67,144 @@ namespace Fortran {
+
+  llvm::SmallVector<mlir::Value> bounds;
+  genBoundsOps(declareOp.getOriginalBase(), bounds);
+  mlir::omp::ClauseMapFlags parentMapFlag = mlir::omp::ClauseMapFlags::to;
```
Done.

TIFitis commented on flang/lib/Lower/OpenMP/Utils.cpp:
```diff
@@ -61,6 +67,144 @@ namespace Fortran {
+                                            /*uniq_name=*/"");
+
+  const auto genBoundsOps = [&](mlir::Value mapVal,
+                                llvm::SmallVectorImpl<mlir::Value> &bounds) {
```
I have a PR for it #158658. But it's been neglected, it needs tweaks to the CMake files etc to work properly and not crash the build bots.

I'll try to get back to it soon.

agozillon replied:
Ah that's where I remember it from! Thank you for working on that, and no rush just good to know it's something in the works :-)

TIFitis commented:
> The PR LGTM so far, just a few questions and a NIT or two. And a request for one or two offload tests in check-offload if at all possible, just so we can see if there's any runtime regressions.

Thanks for the review. I've addressed the nits below. I've added a new test `offload/test/offloading/fortran/implicit-derived-enter-exit.f90`, which along with `offload/test/offloading/fortran/explicit-and-implicit-record-field-mapping.f90` should cover this PR.

> I was also wondering if this work would allow us to remove this segment of code: https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp#L948
> 
> I did test removing it and keeping just your PR in with the following example:
> 
> program main implicit none integer:: i
> 
> ```
> type :: bottom_layer
> real(4) :: i
> integer, allocatable :: scalar_i 
> integer(4) :: array_i(10)
> integer, allocatable :: array_k(:)
> integer(4) :: k
> end type bottom_layer
> 
> type :: one_layer
> real(4) :: i
> integer, allocatable :: scalar 
> integer(4) :: array_i(10)
> real(4) :: j
> integer, allocatable :: array_j(:)
> integer(4) :: k
> type(bottom_layer) :: nest
> end type one_layer
> 
> type(one_layer) :: one_l
> 
> allocate(one_l%nest%array_k(10))
> allocate(one_l%array_j(10))
> allocate(one_l%nest%scalar_i) 
> allocate(one_l%scalar)
> 
> do i = 1, 10
>     one_l%nest%array_i(i) = i
> end do
> 
> one_l%nest%scalar_i = 50
> one_l%scalar = 25
> 
> !$omp target
> do i = 1, 10
>     one_l%nest%array_k(i) = one_l%nest%array_i(i) + i + one_l%nest%scalar_i + one_l%scalar
>     one_l%array_j(i) = one_l%nest%array_k(i)
> end do
> 
> !$omp end target
> ```
> 
> print *, one_l%nest%array_k print *, one_l%nest%array_i print *, one_l%array_j end program
> 
> Unfortunately, it seems that the MapInfoFinalization segment is still required for it, as there's unfortunately no declare mappers generated in this case (from what I could see when testing at least). Is this case something we could perhaps support with the declare mapper approach so we could unify the handling of implicit derived type member mapping? Not sure if it's a trivial extension or something that'd require a fair bit of legwork, if it's the latter we can likely leave it for now :-)

The declare-mapper path only kicks in when the symbol being mapped has an explicit derived-type spec we can inspect during clause processing. ClauseProcessor never materialises the nested-object list that `requiresImplicitDefaultDeclareMapper` needs, so we don’t emit a mapper and rely on the finaliser to materialise the descriptor chain. Supporting that via declare mappers would require us to synthesise implicit mapper ops for every derived symbol captured during region outlining (so they exist even without clauses) and to make the runtime aware of them. That’s a larger follow-up I can work on in the near future.

TIFitis committed:
Address reviewer comments.
